rfcs/0000_code_organization_guide.md

```diff
@@ -0,0 +1,555 @@
+- Start Date: 2025-07-17
+- RFC PR: (leave this empty, it will be filled in after RFC is merged)
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [BCP 14](https://www.rfc-editor.org/info/bcp14), [IETF RFC 2119](https://www.rfc-editor.org/rfc/rfc2119), [IETF RFC 8174](https://www.rfc-editor.org/rfc/rfc8174) when, and only when, they appear in all capitals, as shown here.
+
+This guidance does not dictate the only way to implement functionality. There are many ways to implement any particular functionality, each of which will work. This guidance _**establishes constraints**_ so that as more functionality is implemented and functionality is changed, it remains as easy as it was with the first piece of functionality.
+
+Please note that there are differences between research and platform requirements. While research needs speed and agility, the platform needs modularity and stability. These are different and can conflict. The guidance here is for the platform. _**If you are a researcher, you MAY ignore this guidance in your prototype and code in the way most effective for you and your task.**_ Later, if your prototype works and needs to be integrated into Monty, then we will refactor the prototype to correspond to the guidance here.
+
+# Guidance
+
+## Abstract classes MAY be used to specify interfaces without implementations:
+
+_Why?_ We want to move to [Protocols](https://typing.python.org/en/latest/spec/protocol.html#protocols) eventually. Keeping the abstract classes free of implementation makes it easier to transition to Protocols in the future. Whereas, having an abstract class with some implementation requires additional refactoring when we transition to Protocols.
+
+_Why do we want to move to Protocols eventually?_ We want to catch errors as early as possible, and using Protocols allows us to do this at type check time. Using abstract classes delays this until class instantiation, once the program runs.
+
+```python
+# ABCs raises errors during instantiation when/if constructor is called:
+class Monty(ABC):
+	def implemented(self):
+		pass
+
+	@abstractmethod
+	def unimplemented(self):
+		pass
+
+class DefaultMonty(Monty):
+	pass
+
+monty= DefaultMonty() # runtime error
+
+# ---
+
+# Typical inheritance raises errors during runtime when monty.unimplemented() is called:
+class Monty:
+	def implemented(self):
+		pass
+
+	def unimplemented(self):
+		raise NotImplementedError
+
+class DefaultMonty(Monty):
+	pass
+
+monty = DefaultMonty() # ok
+monty.unimplemented() # runtime error
+
+# ---
+
+# Protocols raise errors during type check when monty.unimplemented() is written:
+class MontyProtocol(Protocol):
+	def implemented(self): ...
+	def unimplemented(self): ...
+
+class DefaultMonty:
+	def implemented(self):
+		pass
+
+monty: MontyProtocol = DefaultMonty() # ok
+monty.unimplemented() # fails type check
```

Comment:
Sorry for the naive question but why wouldn't this fail the typecheck without the protocol? Shouldn't this also raise a type error:
```
class DefaultMonty:
	def implemented(self):
		pass

monty = DefaultMonty()
monty.unimplemented()
```
Why does the typechecker not check whether `DefaultMonty` adheres to `MontyProtocol` when we call `monty: MontyProtocol = DefaultMonty()`?

Reply:
I tested this and when I cpy the protocol code example mypy already raises and error when initializing the DefaultMonty class (which is what I would want the protocol to help with). When I use the code without the protocol I also get a type error on the line where unimplemented is called.